### PR TITLE
#20972: Pre-release QA

### PR DESCRIPTION
**Pull request by bctiemann**
Follow-up QA for the channelized-subinterface work introduced under #20972.

### Changes

* Renaming a channelized parent interface now updates the names of any channel subinterfaces which follow the `<parent name>:<channel ID>` convention, so they stay consistent with the new parent name. A child named otherwise, or whose renamed form would collide with an existing sibling on the same device, is left untouched.
* Relaxed the channel-type restriction so `channel_id` may be assigned to any physical interface type, not just the generic `channel` type. This lets a channel subinterface record its actual transceiver (e.g. `10gbase-sr`) instead of being forced into the generic placeholder — the alternative the original feature request itself called out. Virtual, wireless, LAG, and bridge types still cannot carry a `channel_id`, and a physical type still cannot be assigned a `parent` without one.
* `Interface.is_wired` now also checks `channel_id`, so a channel subinterface using an otherwise-connectable physical type still correctly reports as not directly cable-able.
* Updated `docs/models/dcim/interface.md` to describe both behaviors.
* Added tests covering the rename cascade (including the non-conforming-name and collision cases) and the relaxed type validation (including the still-rejected virtual/LAG cases and `is_wired`).

### CODE REVIEW

**Comment by bctiemann** [comment]
Addressed the two specific suggestions from local review:

**`is_channel` was stale** — it still checked the literal `type == TYPE_CHANNEL`, which disagreed with `is_wired`'s (correct) definition of a channel subinterface as "has `channel_id` set," once a channel subinterface can keep its own specific physical type. Updated `is_channel` to check `self.channel_id is not None`, so the two properties agree.

**`InterfaceTemplate` had the identical latent rename bug** — extended the rename cascade to `InterfaceTemplate`, mirroring the `Interface` fix: renaming a channelized parent template now updates conforming channel subinterface template names (`<parent>:<channel_id>`), skipping non-conforming names and any collision. Collision checks are scoped to the same `device_type` or `module_type`, matching `InterfaceTemplate`'s actual uniqueness constraints (which are per-device-type-or-module-type, not a single combined constraint) — verified this scoping is correct given `device_type`/`module_type` are mutually exclusive per `ModularComponentTemplateModel.clean()`.

Added 7 new tests (4 for the template rename cascade, including a cross-device-type collision non-interference case; 3 for `is_channel`), and verified the template-rename tests actually fail without the fix before confirming they pass with it.

Not addressing suggestions #3 (check-then-write race in the collision guard) or #4 (bulk-rename ordering interaction) from the review — those are narrower, harder-to-hit edge cases rather than the two you called out as worth fixing now.

All 47 tests in `dcim.tests.test_channelization` pass, plus 2332 across `test_models`/`test_api`/`test_filtersets` with no regressions. Ruff clean. Pushed in `13c195ede`.

**Comment by bctiemann** [comment]
Addressed suggestions #3 and #4 too.

**#3 (check-then-write race)** — replaced the Python-level `.exists()` pre-check + single `bulk_update()` with per-child renames, each in its own savepoint (`transaction.atomic()`), catching `IntegrityError`. The database's unique constraint is now the sole arbiter of a collision, closing the TOCTOU window, and — as a bonus correctness improvement — a collision on one child can no longer abort the rename of unrelated siblings in the same call (a single-statement `bulk_update()` covering multiple rows fails atomically for all of them if any one violates a constraint).

**#4 (bulk-rename ordering interaction)** — this turned out to be a real, reachable bug under the *default* queryset ordering, not just an exotic edge case. `Interface`'s default ordering (`device`, natural-sorted name) puts a channelized parent *before* its channel children, so in `BulkRenameView`'s per-object save loop, the parent's cascade fires first — and if the child is also selected in the same batch with a pattern that doesn't happen to match its own name (e.g. a regex-anchored exact match intended for the parent only), the child's own subsequent `save()` uses a stale in-memory copy fetched before the cascade ran, silently overwriting the rename back to the old name. Fixed by deferring the cascade via `transaction.on_commit()`, so it runs after all of the batch's other saves have completed and observes final state rather than a mid-batch snapshot.

Verified both fixes:
- Confirmed each new test fails without its corresponding fix, then restored the fix and confirmed it passes (same revert/restore discipline as the earlier rounds).
- Verified #4 against a **live `BulkRenameView` request** (not just the unit test) — POSTed a real bulk-rename with `find='^et0$', replace='et1', use_regex=True` selecting both a channelized parent and its channel subinterface: final state is `parent.name = 'et1'`, `child.name = 'et1:1'` — the cascade wins, as intended.

Both fixes apply identically to `Interface` and `InterfaceTemplate`. All 51 tests in `dcim.tests.test_channelization` pass, plus 2332 across `test_models`/`test_api`/`test_filtersets` with no regressions. Ruff clean. Pushed in `192cdd2cf`.

**Comment by bctiemann** [comment]
@claude /review - limit response to items of concern only.

**Comment by bctiemann** [comment]
Fixed all findings except 7.

**Comment by bctiemann** [comment]
@claude /review items addressed since last review, with same guardrails.

**Comment by bctiemann** [comment]
Fixed both High and Medium findings from this round.

**High (update() → save() switch re-enables a per-child cable-path rebuild) — fixed and confirmed severe.** Measured it directly before fixing: renaming a 4-channel cabled parent issued **830 queries**. The root cause is exactly as described — `update_channelized_cable_paths` (the `post_save` receiver on `Interface`) re-derives channelization state unconditionally, and a channel subinterface always has `channel_id` set, so every child `save()` during the rename re-ran a full `propagate_channel_cables()` + `rebuild_cable_paths()` for the shared cabled parent. Fixed by scoping the child's `save()` to `update_fields=['name', '_name', 'last_updated']` and adding an early-return in the receiver when `update_fields` is given and touches none of the fields it actually cares about. After the fix: **117 queries** for the same scenario.

While implementing this, the full regression suite caught a real bug in my first pass: `Module`'s manual `post_save.send()` call (for `bulk_update()`-adopted components) passes `update_fields` as a plain `list`, not a `frozenset` — my guard's `.isdisjoint()` call crashed on it. Fixed by using a set-intersection check instead, which works on any iterable.

**Medium (`update_fields` ignored by rename detection) — fixed.** `_detect_channel_rename()`/`_defer_channel_rename()` now thread `update_fields` through from `Interface.save()`/`InterfaceTemplate.save()`. A save with `update_fields` excluding `'name'` no longer cascades the (unpersisted) rename to children, and no longer refreshes `_original_name` to the unpersisted in-memory value — a later full save still correctly detects the rename against the true prior name.

Added a test comparing query cost for a 2-channel vs. 8-channel cabled parent rename (rather than an absolute threshold, which would be fragile against unrelated future query-count drift elsewhere) — confirmed it fails with **2978 queries** on the reverted code and passes on the fix. Added two regression tests for the `update_fields` fix, each confirmed to fail without it.

**Comment by bctiemann** [comment]
@claude /review items fixed in last commit.

File: netbox/dcim/models/mixins.py [hunk 1]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
+    """
+
+    def _init_channel_rename_tracking(self):
+        self._original_name = self.__dict__.get('name')
+
+    def _detect_channel_rename(self, update_fields=None):
+        # A save() whose update_fields excludes 'name' won't actually persist self.name, so neither the cascade
+        # nor the _original_name refresh below may treat it as current.
+        name_persisted = update_fields is None or 'name' in update_fields
+        renamed = bool(self.pk and self.channels and name_persisted and self.name != self._original_name)
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we use the persisted channel state here when `channels` is excluded from `update_fields`?

For example, setting `self.channels = None` and saving only `name` leaves the parent channelized in the database, but makes `renamed` false here. The parent name is persisted and `_original_name` is refreshed, while the children keep their old names.

A regression test for both `Interface` and `InterfaceTemplate` would be helpful.

File: netbox/dcim/models/mixins.py [hunk 2]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
 
 
+class ChannelRenameMixin:
+    """
+    Shared save()-time logic for Interface and InterfaceTemplate: detects a rename of a channelized parent and
+    cascades it to any channel subinterface which follows the "<parent name>:<channel ID>" naming convention.
+
+    A model using this mixin must call _init_channel_rename_tracking() from __init__() (after super().__init__()),
+    then _detect_channel_rename(update_fields) from save() before calling super().save(), and finally
+    _defer_channel_rename(*that result) from save() after calling super().save().
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could this mixin own its lifecycle through cooperative `__init__()` and `save()` methods?

Requiring each model to call three private helpers in a fixed order leaks the implementation into both consumers and makes the mixin easy to wire incorrectly. It would also avoid passing a four-item positional state tuple between the helpers.

File: netbox/dcim/models/mixins.py [hunk 3]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
+
+    def _defer_channel_rename(self, renamed, old_name, new_name, name_persisted):
+        if name_persisted:
+            self._original_name = self.name
+
+        if renamed:
+            # Deferred until commit, not run inline, so it sees the final post-transaction state rather than a
+            # mid-batch snapshot: BulkRenameView fetches every selected object up front and saves each in turn
+            # inside one atomic() block, so a channel subinterface selected in the same batch would otherwise
+            # have its own (stale) save() silently undo the rename applied here moments earlier.
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we shorten this to the invariant rather than documenting `BulkRenameView` internals?

Something like “Defer until commit so a later save in the same transaction cannot overwrite the cascade” should be enough here.

File: netbox/dcim/models/mixins.py [hunk 4]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
+            transaction.on_commit(
+                lambda: self.rename_channel_subinterfaces(old_name, new_name),
+                using=router.db_for_write(type(self)),
+            )
+
+    def rename_channel_subinterfaces(self, old_name, new_name):
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: This is only used as an internal `on_commit()` callback. Could we keep it private as `_rename_channel_subinterfaces()`?

File: netbox/dcim/signals.py [hunk 5]
```diff
@@ -185,15 +185,24 @@ def nullify_connected_endpoints(instance, **kwargs):
...
     Rebuild cable paths when an interface's channelization changes without the Cable itself being modified: a channel
     subinterface is added, moved between parents, or has its channel_id changed, or channelization is toggled on an
     already-cabled interface. (The cable-tracing signals only fire when a Cable is saved.)
     """
     if raw:
         return
+    # update_fields may be a plain list rather than a frozenset (e.g. from Module's manual post_save.send()),
+    # so isdisjoint() can't be relied on directly.
+    if update_fields is not None and not (set(update_fields) & CHANNELIZATION_RELEVANT_FIELDS):
+        return
```
Comment (pheus) [inline netbox/dcim/signals.py]: Could we make this receiver-specific constant private and simplify the check? `frozenset.isdisjoint()` accepts a list, so the `set()` allocation and the explanation above are unnecessary:

```python
if update_fields is not None and _CHANNELIZATION_RELEVANT_FIELDS.isdisjoint(update_fields):
    return
```

File: docs/models/dcim/interface.md [hunk 6]
```diff
@@ -84,14 +84,14 @@ If selected, this component will be treated as if a cable has been connected.
 
 ### Parent Interface
 
-Virtual interfaces can be bound to a physical parent interface. This is helpful for modeling virtual interfaces which employ encapsulation on a physical interface, such as an 802.1Q VLAN-tagged subinterface. Channel-type subinterfaces are likewise bound to their [channelized](#channels) parent interface.
+Virtual interfaces can be bound to a physical parent interface. This is helpful for modeling virtual interfaces which employ encapsulation on a physical interface, such as an 802.1Q VLAN-tagged subinterface. A channel subinterface is likewise bound to its [channelized](#channels) parent interface, whether it uses the generic **channel** type or its own specific physical type.
 
 !!! note
-    An interface with one or more child interfaces assigned cannot be deleted until all its child interfaces have been deleted or reassigned.
+    An interface with one or more child interfaces assigned cannot be deleted until all its child interfaces have been deleted or reassigned. Renaming a channelized interface updates the names of any channel subinterfaces which follow the `<name>:<channel ID>` convention, to keep their names consistent with their new parent.
```
Comment (pheus) [inline docs/models/dcim/interface.md]: Could we mention the collision exception here?

A conforming child can remain unchanged when the generated name already exists, so the current wording is stronger than the implementation.

File: netbox/dcim/models/mixins.py [hunk 7]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
+        for child in self.child_interfaces.filter(channel_id__isnull=False):
+            if child.name != f'{old_name}:{child.channel_id}':
+                continue
+            # A full save() (not a queryset update()) so _name, last_updated, and the changelog get updated too;
+            # a channel subinterface can never itself be channelized, so this can't recurse into the cascade.
+            # update_fields is restricted to what actually changed so unrelated receivers (e.g. Interface's own
+            # cable-path rebuild) can skip redundant work.
+            if hasattr(child, 'snapshot'):
+                child.snapshot()
+            child.name = f'{new_name}:{child.channel_id}'
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we guard the generated name length before saving?

A 62-character parent can have a valid `<parent>:1` child, but renaming the parent to 64 characters produces a 66-character child name. Because this runs after commit and the handler below catches only `IntegrityError`, the parent can remain renamed while the child keeps its old name.

Please cover this boundary for both models.

File: netbox/dcim/models/mixins.py [hunk 8]
```diff
@@ -242,6 +257,65 @@ def clean(self):
...
+            # Renamed in its own savepoint: the DB's unique constraint is the sole arbiter of a collision, and a
+            # collision on one child can't abort the rename of the others.
+            try:
+                with transaction.atomic(using=router.db_for_write(type(self))):
+                    child.save(update_fields=['name', '_name', 'last_updated'])
+            except IntegrityError:
+                continue
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we avoid swallowing every `IntegrityError` here?

The save can fail for a reason other than the expected name collision. Please confirm that the generated sibling name now exists before continuing, and re-raise otherwise.

File: netbox/dcim/tests/test_channelization.py [hunk 9]
```diff
@@ -453,6 +569,422 @@ def test_parent_channel_id_must_be_unique(self):
             duplicate.full_clean()
 
 
+class ChannelizedInterfaceRenameTestCase(TestCase):
```
Comment (pheus) [inline netbox/dcim/tests/test_channelization.py]: Could we consolidate the ordinary model-focused test cases?

NetBox generally keeps one model-attached `TestCase` per model. Please fold the `Interface` validation, property, rename, and filter coverage into a single `ChannelizedInterfaceTestCase`, and merge the template rename coverage into `ChannelizedInterfaceTemplateTestCase`.

The cable-path and view tests can remain specialized, and the commit-dependent cases can remain in the separate `TransactionTestCase`.

File: netbox/dcim/tests/test_channelization.py [hunk 10]
```diff
@@ -372,14 +433,69 @@ def test_channel_type_requires_channel_id(self):
         with self.assertRaises(ValidationError):
             interface.full_clean()
 
-    def test_channel_id_requires_channel_type(self):
+    def test_channel_id_allowed_on_specific_physical_type(self):
```
Comment (pheus) [inline netbox/dcim/tests/test_channelization.py:521]: Could we add the equivalent positive case for `InterfaceTemplate`?

The existing template test still describes the old type restriction and now raises because no parent is assigned, so it no longer proves that a physical-type template with a valid channelized parent is accepted.

File: netbox/dcim/tests/test_channelization.py [hunk 11]
```diff
@@ -342,6 +354,55 @@ def test_110_move_channel_to_uncabled_parent(self):
...
+
+    def test_111_rename_cabled_parent_preserves_cable_paths_without_quadratic_cost(self):
+        """
+        Renaming a cabled channelized parent must cascade the children's names without disturbing their cable
+        paths, and without re-deriving cable state per child (which would make the rename quadratic in the
+        channel count). Pinned by comparing query cost at 2 vs. 8 channels: linear per-child work scales with
+        the 4x channel growth; a quadratic regression would blow well past it.
+        """
+        queries_2ch, near_channels, far_channels, cable = self._rename_cabled_channelized_pair('A', 2)
+        queries_8ch, _, _, _ = self._rename_cabled_channelized_pair('B', 8)
```
Comment (pheus) [inline netbox/dcim/tests/test_channelization.py]: Could we retain and verify the 8-channel result as well?

That is the case exercising the scaling path, but its objects are discarded, so the test could still pass if the larger cascade did not rename or preserve the paths correctly.

File: docs/models/dcim/interface.md
Comment (pheus) [inline docs/models/dcim/interface.md]: Could we change “channel-type subinterface” to “channel subinterface”?

A channel may now retain its specific physical type, so this sentence still describes the old restriction.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for the thorough follow-up. I’ve left a few inline comments, including two blocking issues with rename propagation and some concerns about complexity and test organization. Requesting changes for now.

File: netbox/dcim/models/mixins.py (hunk 1 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: `is_channelized = self.channels if channels_persisted else self._original_channels` now, so a plain rename doesn't skip the cascade.

File: netbox/dcim/models/mixins.py (hunk 2 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: Fixed to reduce redundancy, and updated the docstring to describe the correct use of the MRO. Removed `InterfaceTemplate.__init__/save()` entirely.

File: netbox/dcim/signals.py (hunk 5 above)
Comment (bctiemann) [reply netbox/dcim/signals.py]: Done. `isdisjoint()` threw an error in my local test development iterations, but local Claude and GitHub Claude got increasingly strident in their disagreements about it

File: netbox/dcim/models/mixins.py (hunk 7 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: We now skip a name exceeding the field length before attempting save.

File: netbox/dcim/models/mixins.py (hunk 8 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: Now calls `validate_constraints()` to turn it into a `ValidationError` if possible.

File: netbox/dcim/tests/test_channelization.py (hunk 9 above)
Comment (bctiemann) [reply netbox/dcim/tests/test_channelization.py]: Merged all functional tests into `ChannelizedInterfaceTestCase` and template tests into `ChannelizedInterfaceTemplateTestCase`

File: netbox/dcim/tests/test_channelization.py (hunk 10 above)
Comment (bctiemann) [reply netbox/dcim/tests/test_channelization.py:521]: Added `test_template_channel_id_allowed_on_specific_physical_type`

File: netbox/dcim/tests/test_channelization.py (hunk 11 above)
Comment (bctiemann) [reply netbox/dcim/tests/test_channelization.py]: The perf test now asserts correctness for both the 2- and 8-channel cascades, not just the discarded larger one.

File: netbox/dcim/models/mixins.py [hunk 12]
```diff
@@ -242,6 +257,86 @@ def clean(self):
...
+            # a channel subinterface can never itself be channelized, so this can't recurse into the cascade.
+            # update_fields is restricted to what actually changed so unrelated receivers (e.g. Interface's own
+            # cable-path rebuild) can skip redundant work.
+            if hasattr(child, 'snapshot'):
+                child.snapshot()
+            child.name = candidate_name
+            # Renamed in its own savepoint: the DB's unique constraint is the sole arbiter of a collision, and a
+            # collision on one child can't abort the rename of the others.
+            try:
+                with transaction.atomic(using=router.db_for_write(type(self))):
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we wrap the whole cascade in one outer `atomic()` and keep this block nested?

The callback runs after the original transaction has committed, so this is currently a separate transaction per child rather than a savepoint. An unexpected failure can leave only part of the child set renamed.

File: netbox/dcim/models/mixins.py [hunk 13]
```diff
@@ -242,6 +257,86 @@ def clean(self):
             raise ValidationError({'rf_role': _("Wireless role may be set only on wireless interfaces.")})
 
 
+class ChannelRenameMixin:
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we call this `InterfaceChannelRenameMixin`? The behavior is specific to `Interface` and `InterfaceTemplate`, and the narrower name matches `InterfaceValidationMixin`.

File: netbox/dcim/models/device_component_templates.py [hunk 14]
```diff
@@ -663,12 +663,8 @@ class Meta(ModularComponentTemplateModel.Meta):
...
         verbose_name_plural = _('interface templates')
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        # Cache the original channel count for use by InterfaceValidationMixin.clean() to detect a channel-count
-        # reduction that would orphan a bound subinterface.
-        self._original_channels = self.__dict__.get('channels')
+    # __init__() and save() are entirely handled by ChannelRenameMixin (first in the MRO above); no override
+    # needed here. _original_channels, set there, is also used below by InterfaceValidationMixin.clean().
```
Comment (pheus) [inline netbox/dcim/models/device_component_templates.py]: Could we drop this comment? The mixin already documents its lifecycle, and we don't need to explain why an override is absent.

File: docs/models/dcim/interface.md [hunk 15]
```diff
@@ -84,14 +84,14 @@ If selected, this component will be treated as if a cable has been connected.
...
+Virtual interfaces can be bound to a physical parent interface. This is helpful for modeling virtual interfaces which employ encapsulation on a physical interface, such as an 802.1Q VLAN-tagged subinterface. A channel subinterface is likewise bound to its [channelized](#channels) parent interface, whether it uses the generic **channel** type or its own specific physical type.
 
 !!! note
-    An interface with one or more child interfaces assigned cannot be deleted until all its child interfaces have been deleted or reassigned.
+    An interface with one or more child interfaces assigned cannot be deleted until all its child interfaces have been deleted or reassigned. Renaming a channelized interface updates the names of any channel subinterfaces which follow the `<name>:<channel ID>` convention, to keep their names consistent with their new parent, unless the resulting name is already in use by another interface on the device (in which case that subinterface's name is left unchanged).
```
Comment (pheus) [inline docs/models/dcim/interface.md]: Could we mention the length case here as well? The implementation also leaves the child unchanged when the generated name exceeds the field limit.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for the update. Most of the earlier points are addressed. I’ve left one blocking transaction issue and a few small cleanup items. Requesting changes for now.

File: docs/models/dcim/interface.md (hunk 15 above)
Comment (bctiemann) [reply docs/models/dcim/interface.md]: Addressed in `interface.md`

File: netbox/dcim/models/mixins.py [hunk 16]
```diff
@@ -242,6 +257,93 @@ def clean(self):
...
+        if name_persisted:
+            self._original_name = new_name
+        if channels_persisted:
+            self._original_channels = self.channels
+
+        if renamed:
+            # Defer until commit so a later save in the same transaction cannot overwrite the cascade.
+            transaction.on_commit(
+                lambda: self._rename_channel_subinterfaces(old_name, new_name),
+                using=router.db_for_write(type(self)),
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Could we capture `self._state.db` after `super().save()` and pass it through the callback, child query, and saves? Re-running the router here can differ from an explicit `save(using=...)`.

File: netbox/dcim/models/mixins.py [hunk 17]
```diff
@@ -168,21 +182,23 @@ def clean(self):
...
         # The channel type and channel_id are mutually dependent. The channel_id requirement is relaxed for a
         # replication base (bulk creation), where each channel_id is supplied per-instance during expansion.
-        is_channel = self.type == InterfaceTypeChoices.TYPE_CHANNEL
-        if is_channel and self.channel_id is None and not getattr(self, '_replicated_base', False):
+        if self.type == InterfaceTypeChoices.TYPE_CHANNEL and self.channel_id is None and not is_replicated_base:
             raise ValidationError({
                 'channel_id': _("Channel interfaces must have a channel ID assigned.")
             })
-        if self.channel_id is not None and not is_channel:
+        if self.channel_id is not None and not can_bind_to_channel:
```
Comment (pheus) [inline netbox/dcim/models/mixins.py:189]: Could we reject assigning a channel to an interface with an existing `CableTermination`? This transition is currently accepted, but the signal then clears or replaces the cached cable fields while the original termination remains.

I think this belongs in `Interface.clean()` and should check `cable_terminations`, not `self.cable`, since a valid channel child intentionally mirrors its parent's cable.

File: netbox/dcim/signals.py
Comment (pheus) [inline netbox/dcim/signals.py]: Could we gate this on creation or an actual change to `parent_id` or `channel_id`? The `update_fields` guard handles partial saves, but a full save of an existing channel child still propagates cable state and rebuilds the parent's paths for unrelated changes.

Comment (pheus) [inline netbox/dcim/signals.py]: Could we clear the mirrored cable fields when a channel child is detached? Clearing both `parent` and `channel_id` skips this branch, and the former child is no longer included in the old parent's propagation queryset, so it retains the old cable cache.

A regression test for this transition would be helpful.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for the follow-up. I found a few remaining correctness and performance issues around channel handling and left them inline. Requesting changes for now.

File: netbox/dcim/models/mixins.py (hunk 16 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: Done.

File: netbox/dcim/models/mixins.py (hunk 17 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py:189]: Done.

File: netbox/dcim/signals.py
Comment (bctiemann) [reply netbox/dcim/signals.py]: `update_channelized_cable_paths()` now only propagates/rebuilds when the save is a creation or an actual `channel_id`/`parent_id`.

Comment (bctiemann) [reply netbox/dcim/signals.py]: Done.

Comment (pheus) [inline netbox/dcim/signals.py:267]: Could we keep `instance`'s cable state in sync here as well?

`MACAddressShortcutMixin` saves the same interface again when `mac_address` changes. Since propagation and cleanup update other queryset instances, that second full save can write the stale cable cache back after these `_original_*` values have been refreshed.

Please add an API regression test combining a channel change with `mac_address`.

File: netbox/dcim/models/mixins.py [hunk 18]
```diff
@@ -168,21 +182,37 @@ def clean(self):
...
+                    "A channel ID cannot be assigned to a virtual, LAG, bridge, or wireless interface."
+                )
+            })
+
+        # A channel subinterface's cable state is mirrored from its channelized parent (see
+        # update_channelized_cable_paths()), so it cannot also carry its own CableTermination -- checking
+        # cable_terminations rather than self.cable, since a valid channel child's self.cable is expected to
+        # already reflect the parent's mirrored cable. InterfaceTemplate shares this mixin but is not a
+        # CabledObjectModel and so has no cable_terminations at all.
+        cable_terminations = getattr(self, 'cable_terminations', None)
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Suggestion: Could this check live in `Interface.clean()`? It applies only to `Interface`, and moving it there would avoid the defensive `getattr()` for `InterfaceTemplate`.

File: netbox/dcim/models/mixins.py [hunk 19]
```diff
@@ -242,6 +271,96 @@ def clean(self):
...
+                if child.name != f'{old_name}:{child.channel_id}':
+                    continue
+                candidate_name = f'{new_name}:{child.channel_id}'
+                if len(candidate_name) > max_name_length:
+                    continue
+                # A full save() (not a queryset update()) so _name, last_updated, and the changelog get updated
+                # too; a channel subinterface can never itself be channelized, so this can't recurse into the
+                # cascade. update_fields is restricted to what actually changed so unrelated receivers (e.g.
+                # Interface's own cable-path rebuild) can skip redundant work.
+                if hasattr(child, 'snapshot'):
```
Comment (pheus) [inline netbox/dcim/models/mixins.py]: Suggestion: Both supported models provide `snapshot()`. Calling it directly would make the mixin's contract explicit rather than silently skipping change logging if it is used incorrectly.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks, the earlier issues are fixed. I found one remaining API consistency issue and left two optional suggestions inline. Requesting changes for the first one.

File: netbox/dcim/signals.py
Comment (bctiemann) [reply netbox/dcim/signals.py:267]: Done, and added an API regression test combining a channel-binding PATCH with `mac_address` (verified it fails pre-fix, passes post-fix).

File: netbox/dcim/models/mixins.py (hunk 19 above)
Comment (bctiemann) [reply netbox/dcim/models/mixins.py]: `_rename_channel_subinterfaces()` now calls `child.snapshot()` directly.

File: netbox/dcim/signals.py [hunk 20]
```diff
@@ -225,6 +241,16 @@ def update_channelized_cable_paths(instance, created, raw=False, **kwargs):
...
 
+    # A channel child's own cable_id/cable_end/cable_connector/cable_positions may have just been mutated at the
+    # DB level above -- either mirrored from its parent (propagate_channel_cables(), which bulk_updates a
+    # separately-fetched copy of this same row) or cleared (the queryset .update() above) -- without touching
+    # this in-memory `instance`. A later full save() of this same instance by another caller in the same request
+    # (e.g. MACAddressShortcutMixin.update()'s second instance.save() for a combined mac_address change) would
+    # otherwise write those stale in-memory values back over what was just written. Refresh from the DB so it
+    # doesn't.
+    if instance.channel_id or instance._original_channel_id:
+        instance.refresh_from_db(fields=['cable', 'cable_end', 'cable_connector', 'cable_positions'])
```
Comment (pheus) [inline netbox/dcim/signals.py]: Could we include `_path` in this refresh? `rebuild_cable_paths()` updates it through queryset operations as well, so the MAC shortcut's second full save can still write a stale `_path_id` back.

Please extend the API regression coverage to verify the child's path when attaching and detaching it.

File: netbox/dcim/signals.py [hunk 21]
```diff
@@ -225,6 +241,16 @@ def update_channelized_cable_paths(instance, created, raw=False, **kwargs):
...
                     cablepath.delete()
 
+    # A channel child's own cable_id/cable_end/cable_connector/cable_positions may have just been mutated at the
+    # DB level above -- either mirrored from its parent (propagate_channel_cables(), which bulk_updates a
+    # separately-fetched copy of this same row) or cleared (the queryset .update() above) -- without touching
+    # this in-memory `instance`. A later full save() of this same instance by another caller in the same request
+    # (e.g. MACAddressShortcutMixin.update()'s second instance.save() for a combined mac_address change) would
+    # otherwise write those stale in-memory values back over what was just written. Refresh from the DB so it
+    # doesn't.
+    if instance.channel_id or instance._original_channel_id:
```
Comment (pheus) [inline netbox/dcim/signals.py]: Suggestion: Could we refresh only when propagation or cleanup actually changed this row? This currently adds a query to every full save of a channel child, even when its binding is unchanged.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks, the previous comments are addressed. I found one remaining issue in the same API flow and a small suggestions inline. Requesting changes for the first one.

File: netbox/dcim/signals.py (hunk 20 above)
Comment (bctiemann) [reply netbox/dcim/signals.py]: `update_channelized_cable_paths()` now also refreshes `_path` on the channel child's in-memory instance.

File: netbox/dcim/signals.py (hunk 21 above)
Comment (bctiemann) [reply netbox/dcim/signals.py]: Now gated on an `own_row_mutated` flag set only when this function actually wrote to the instance's own row via a queryset/bulk operation.

**Review by pheus — APPROVED** [review]
Thanks, the latest changes address the remaining issue! This looks good to me!
